h3-webtransport/src/server.rs

```diff
@@ -0,0 +1,422 @@
+
+            tracing::debug!("Client settings: {:#?}", config);
+            if !config.enable_webtransport {
+                return Err(conn.close(
+                    Code::H3_SETTINGS_ERROR,
+                    "webtransport is not supported by client",
+                ));
+            }
+
+            if !config.enable_datagram {
+                return Err(conn.close(
+                    Code::H3_SETTINGS_ERROR,
+                    "datagrams are not supported by client",
+                ));
+            }
+        }
+
+        tracing::debug!("Validated client webtransport support");
+
+        // The peer is responsible for validating our side of the webtransport support.
+        //
+        // However, it is still advantageous to show a log on the server as (attempting) to
+        // establish a WebTransportSession without the proper h3 config is usually a mistake.
+        if !conn.inner.config.enable_webtransport {
+            tracing::warn!("Server does not support webtransport");
+        }
+
+        if !conn.inner.config.enable_datagram {
+            tracing::warn!("Server does not support datagrams");
+        }
+
+        if !conn.inner.config.enable_extended_connect {
+            tracing::warn!("Server does not support CONNECT");
+        }
+
+        // Respond to the CONNECT request.
+
+        //= https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3/#section-3.3
+        let response = if validate_wt_connect(&request) {
+            Response::builder()
+                // This is the only header that chrome cares about.
+                .header("sec-webtransport-http3-draft", "draft02")
+                .status(StatusCode::OK)
+                .body(())
+                .unwrap()
+        } else {
+            Response::builder()
+                .status(StatusCode::BAD_REQUEST)
+                .body(())
+                .unwrap()
+        };
+
+        tracing::info!("Sending response: {response:?}");
+        stream.send_response(response).await?;
+
+        tracing::info!("Stream id: {}", stream.id());
+        let session_id = stream.send_id().into();
+        tracing::info!("Established new WebTransport session with id {session_id:?}");
+        let conn_inner = conn.inner.conn.lock().unwrap();
+        let opener = Mutex::new(conn_inner.opener());
+        drop(conn_inner);
+
+        Ok(Self {
+            session_id,
+            opener,
+            conn: Mutex::new(conn),
+            connect_stream: stream,
+        })
+    }
+
+    /// Receive a datagram from the client
+    pub fn accept_datagram(&self) -> ReadDatagram<C> {
+        ReadDatagram {
+            conn: self.conn.lock().unwrap().inner.conn.clone(),
+        }
+    }
+
+    /// Sends a datagram
+    ///
+    /// TODO: maybe make async. `quinn` does not require an async send
+    pub fn send_datagram(&self, data: impl Buf) -> Result<(), Error> {
+        self.conn
```

Comment:
@seanmonstar we had to expose the h3 connection via Arc<Mutex> what do you think about this?